diff --git a/neutron_fwaas/services/firewall/service_drivers/agents/agents.py b/neutron_fwaas/services/firewall/service_drivers/agents/agents.py
index 3a3a27b45..4ece54b7b 100644
--- a/neutron_fwaas/services/firewall/service_drivers/agents/agents.py
+++ b/neutron_fwaas/services/firewall/service_drivers/agents/agents.py
@@ -185,17 +185,7 @@ class FirewallAgentDriver(driver_api.FirewallDriverDB,
             log_plugin.driver_manager.register_driver(logging_driver.DRIVER)
 
     def is_supported_l2_port(self, port):
-        if port[pb_def.VIF_TYPE] == pb_def.VIF_TYPE_OVS:
-            if not port['port_security_enabled']:
-                return True
-
-            # TODO(annp): remove these lines after we fully support for hybrid
-            # port
-            if not port[pb_def.VIF_DETAILS][pb_def.OVS_HYBRID_PLUG]:
-                return True
-            LOG.warning("Doesn't support hybrid port at the moment")
-        else:
-            LOG.warning("Doesn't support vif type %s", port[pb_def.VIF_TYPE])
+        LOG.warning("Support for l2 ports was removed")
         return False
 
     def is_supported_l3_port(self, port):
